<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xmlns:th="http://www.thymeleaf.org">
<head>
    <meta charset="UTF-8">
    <title>Squirls - Super-Quick Information Content and Random Forest Learning for Splice Variants</title>
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <!-- TODO - Add the CSS into the template -->
    <!-- Latest compiled and minified CSS -->
    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css" rel="stylesheet">
    <style>
        body {
            background-color: #eee;
        }

        h1 {
            font-size: 28px;
            font-weight: bold;
        }

        h2 {
            font-size: 20px;
            font-weight: bold;
        }

        h3 {
            font-size: 16px;
            font-style: italic;
        }

        footer div {
            text-align: center;
            padding: 1em;
        }

        .content {
            background-color: white;
            filter: drop-shadow(5px 5px 5px #aaa);
            padding: 40px
        }

        .underlined {
            border-bottom: thin solid #bce8f1;
        }

        .overlined {
            border-top: thin solid #bce8f1;
        }

        .input-resource {
            font-style: italic;
            background-color: #f9f9f9;
            border-top: solid thin #efefef;
            border-bottom: solid thin #efefef;
            padding: .5em;
        }

        .row {
            display: flex;
            place-content: center;
            padding: 0 1em 0 1em;
        }

        .column {
            float: left;
            width: 50%;
            padding: 10px;
            min-height: 100px;
        }

        /* Clear floats after the columns */
        .row:after {
            content: "";
            display: table;
            clear: both;
        }

        .graphics-container {
            width: 100%;
            height: 100%;
            padding-top: 5px;
        }

        .graphics-title {
            padding-top: .7rem;
            font-weight: bold;
        }

        /* Individual figures are wrapped in a div of this class. */
        .graphics-subcontent {
            text-align: center;
            padding: 2rem;
            display: table-cell;
            vertical-align: middle;
        }

        .donor .graphics-subcontent:nth-child(1) svg:nth-child(1) {
            display: block;
        }

        /* All figures wrapped within `.graphics-subcontent` are wrapped within this element. */
        .graphics-content {
            margin: 0 auto;
            /*padding: 2rem;*/
            display: flex;
            justify-content: center;
        }

        .graphics-label {
            padding: 0 0 .7rem 0;
        }

        div svg {
            object-fit: contain;
        }

        ul.bottomspace {
            margin-bottom: 1cm;
        }

        .interpretation {
            border: solid thin #ffb4b4;
            background-color: #fff4f4;
            padding: 10px;
            border-radius: 5px;
        }

        .variant-section-title {
            font-size: 16px;
            font-weight: bold;
            margin: 5px 0 5px 0;
        }

        .transcript-list {
            margin: 0 0 15px 0;
        }

        .feature-table tr {
            display: block;
            float: left;
        }

        .feature-table th, .feature-table td {
            display: block;
        }

        .feature-table td {
            text-align: center;
            border: solid .5px #ddd;
        }

    </style>
</head>
<body>
  <div class="container content">
      <header class="underlined">
          <h1><em>Squirls</em> analysis results for <em th:text="${sampleName}">Sample ID</em></h1>
          <p>
              <em>Squirls</em> is a desktop Java application that performs prediction
              of deleteriousness of genomic variants with respect to mRNA splicing.
              Variants are prioritized based on interpretable splicing features using
              a machine learning algorithm. The functional annotation code is performed
              using the <a href="https://github.com/charite/jannovar">Jannovar</a>
              library.
          </p>
          <p>
              See <a href="https://squirls.readthedocs.io/en/latest/">the comprehensive
              documentation</a> learn how to setup and run <em>Squirls</em>, and how
              to interpret the results.
          </p>
      </header>

      <section>
          <div class="row">
              <div class="column" id="inputs">
                  <!--    INPUTS    -->
                  <h2>Analysis input</h2>
                  <h3>Analyzed variants</h3>
                  <p>
                      <em>Squirls</em> analyzed variants stored in the VCF file located at:
                  </p>
                  <p><div class="input-resource" th:text="${settings.getInputPath()}">/path/to/file.vcf</div></p>

                  <h3>Transcript definitions</h3>
                  <p>
                      <em>Squirls</em> used <em><span th:remove="tag" th:text="${settings.getFeatureSource()}">RefSeq</span></em> genes
                      and evaluated the variants with respect to transcripts from the <em><span th:remove="tag" th:text="${settings.getTranscriptCategory()}">Manual</span></em> transcript category:
                      <table class="table table-striped">
                          <tbody>
                            <tr th:each="definition: ${settings.definitions()}">
                                <td><span th:remove="tag" th:text="${definition}">Known transcripts (NM_, NR_)</span></td>
                            </tr>
                          </tbody>
                      </table>
                  </p>

                  <h3>Number of reported variants</h3>
                  <p>
                      After the analysis, the variants are sorted by predicted
                      pathogenicity and the variant with the highest predicted
                      pathogenicity is placed on the top of the list.
                      This report presents the top
                      <b><span th:text="${settings.getNReported()}">100</span></b>
                      most pathogenic variants.
                  </p>
              </div>
              <div class="column" id="statistics">
                  <!--    STATISTICS      -->
                  <h2>Statistics</h2>

                  <table class="table table-striped">
                      <tr>
                          <td>Number of variants in the VCF file</td>
                          <td th:text="${stats.allVariants()}">100</td>
                      </tr>
                      <tr>
                          <td>Number of distinct <em>ALT</em> alleles in the VCF file</td>
                          <td th:text="${stats.alleleCount()}">120</td>
                      </tr>
                      <tr>
                          <td>Number of <em>ALT</em> alleles annotated by Squirls</td>
                          <td th:text="${stats.annotatedAlleleCount()}">110</td>
                      </tr>
                      <tr>
                          <td>Number of reported variants</td>
                          <td th:text="${variants.size()}">100</td>
                      </tr>
                  </table>
              </div>
          </div>
      </section>

      <section class="overlined" id="prioritisation">
        <!--    PRIORITISED VARIANTS     -->
        <h2>Prioritised Variants</h2>
        <div class="panel panel-info" th:each="variant, variant_is: ${variants}">
            <div class="panel-heading">
                <div class="panel-title row">
                    <div class="col-md-4">
                        <b th:text="${variant.getVariantRepresentation()}">chr9:136,224,690A>G</b>
                    </div>
                    <div class="col-md-4">
                        <b><em th:text="${variant.getGeneSymbol()}">SURF2</em></b>
                    </div>
                    <div class="col-md-4">
                        Squirls score:
                        <b th:text="${#numbers.formatDecimal(variant.getMaxPathogenicity(), 1, 3)}">0.89</b>
                    </div>
                </div>
            </div>
            <div class="panel-body">
                <div th:remove="tag" th:utext="${variant.getContent()}">
                    <!-- CONTENT  -->
                </div>
            </div>
        </div>
      </section>

      <footer>
          <div>
              <p>Developed by <em>Daniel Danis</em> and <em>Peter N Robinson</em> at <a href="https://www.jax.org">The Jackson Laboratory for Genomic Medicine</a>.</p>
              <p>Problems, suggestions, or comments? Feel free to submit an issue to our <a href="https://github.com/TheJacksonLaboratory/Squirls/issues">GitHub tracker</a>.</p>
              <p>&#169; 2021 Daniel Danis, Peter N Robinson</p>
          </div>
      </footer>
  </div>
</body>
</html>
